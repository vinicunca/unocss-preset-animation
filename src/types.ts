import type { IconsOptions } from '@unocss/preset-icons';
import type { ThemeAnimation } from '@unocss/preset-mini';
import type { PresetUnoOptions } from '@unocss/preset-uno';
import type { WebFontsOptions } from '@unocss/preset-web-fonts';
import type { Preset } from 'unocss';

export interface VinicuncaOptions {
  /**
   * Enable animations
   *
   * @default true
   */
  enableAnimations?: boolean;

  /**
   * Enable default shortcuts
   *
   * @default true
   */
  enableDefaultShortcuts?: boolean;

  /**
   * Enable reset styles
   *
   * @default true
   */
  enableResetStyles?: boolean;

  /**
   * Enable icons preset and the options of it
   * @default true
   */
  icons?: IconsOptions | boolean;

  /**
   * Enable the default preset
   * @default true
   */
  uno?: PresetUnoOptions | boolean;

  /**
   * Enable webFonts preset and the options of it
   * @default false
   */
  webFonts?: WebFontsOptions | boolean;
}

export type ResolvedOptions = {
  animationTheme: ThemeAnimation;
  presets: Array<Preset>;
  // shortcuts: CustomStaticShortcuts;
} & Required<VinicuncaOptions>;
