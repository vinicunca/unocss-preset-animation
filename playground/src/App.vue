<script setup lang="ts">
import { nextTick, ref } from 'vue';

const classLists = [
  'animate-back-in-down',
  'animate-back-in-left',
  'animate-back-in-right',
  'animate-back-in-up',
  'animate-back-out-down',
  'animate-back-out-left',
  'animate-back-out-right',
  'animate-back-out-up',
  'animate-backInDown',
  'animate-backInLeft',
  'animate-backInRight',
  'animate-backInUp',
  'animate-backOutDown',
  'animate-backOutLeft',
  'animate-backOutRight',
  'animate-backOutUp',
  'animate-boingInUp',
  'animate-boingOutDown',
  'animate-bombLeftOut',
  'animate-bombRightOut',
  'animate-bounce',
  'animate-bounce-alt',
  'animate-bounce-in',
  'animate-bounce-in-down',
  'animate-bounce-in-left',
  'animate-bounce-in-right',
  'animate-bounce-in-up',
  'animate-bounce-out',
  'animate-bounce-out-down',
  'animate-bounce-out-left',
  'animate-bounce-out-right',
  'animate-bounce-out-up',
  'animate-bounceIn',
  'animate-bounceInDown',
  'animate-bounceInLeft',
  'animate-bounceInRight',
  'animate-bounceInUp',
  'animate-bounceOut',
  'animate-bounceOutDown',
  'animate-bounceOutLeft',
  'animate-bounceOutRight',
  'animate-bounceOutUp',
  'animate-fade-in',
  'animate-fade-in-bottom-left',
  'animate-fade-in-bottom-right',
  'animate-fade-in-down',
  'animate-fade-in-down-big',
  'animate-fade-in-left',
  'animate-fade-in-left-big',
  'animate-fade-in-right',
  'animate-fade-in-right-big',
  'animate-fade-in-top-left',
  'animate-fade-in-top-right',
  'animate-fade-in-up',
  'animate-fade-in-up-big',
  'animate-fade-out',
  'animate-fade-out-bottom-left',
  'animate-fade-out-bottom-right',
  'animate-fade-out-down',
  'animate-fade-out-down-big',
  'animate-fade-out-left',
  'animate-fade-out-left-big',
  'animate-fade-out-right',
  'animate-fade-out-right-big',
  'animate-fade-out-top-left',
  'animate-fade-out-top-right',
  'animate-fade-out-up',
  'animate-fade-out-up-big',
  'animate-fadeIn',
  'animate-fadeInBottomLeft',
  'animate-fadeInBottomRight',
  'animate-fadeInDown',
  'animate-fadeInDownBig',
  'animate-fadeInLeft',
  'animate-fadeInLeftBig',
  'animate-fadeInRight',
  'animate-fadeInRightBig',
  'animate-fadeInTopLeft',
  'animate-fadeInTopRight',
  'animate-fadeInUp',
  'animate-fadeInUpBig',
  'animate-fadeOut',
  'animate-fadeOutBottomLeft',
  'animate-fadeOutBottomRight',
  'animate-fadeOutDown',
  'animate-fadeOutDownBig',
  'animate-fadeOutLeft',
  'animate-fadeOutLeftBig',
  'animate-fadeOutRight',
  'animate-fadeOutRightBig',
  'animate-fadeOutTopLeft',
  'animate-fadeOutTopRight',
  'animate-fadeOutUp',
  'animate-fadeOutUpBig',
  'animate-flash',
  'animate-flip',
  'animate-flip-in-x',
  'animate-flip-in-y',
  'animate-flip-out-x',
  'animate-flip-out-y',
  'animate-flipInX',
  'animate-flipInY',
  'animate-flipOutX',
  'animate-flipOutY',
  'animate-foolishIn',
  'animate-foolishOut',
  'animate-head-shake',
  'animate-headShake',
  'animate-heart-beat',
  'animate-heartBeat',
  'animate-hinge',
  'animate-holeOut',
  'animate-jack-in-the-box',
  'animate-jackInTheBox',
  'animate-jello',
  'animate-light-speed-in-left',
  'animate-light-speed-in-right',
  'animate-light-speed-out-left',
  'animate-light-speed-out-right',
  'animate-lightSpeedInLeft',
  'animate-lightSpeedInRight',
  'animate-lightSpeedOutLeft',
  'animate-lightSpeedOutRight',
  'animate-magic',
  'animate-openDownLeft',
  'animate-openDownLeftOut',
  'animate-openDownLeftReturn',
  'animate-openDownRight',
  'animate-openDownRightOut',
  'animate-openDownRightReturn',
  'animate-openUpLeft',
  'animate-openUpLeftOut',
  'animate-openUpLeftReturn',
  'animate-openUpRight',
  'animate-openUpRightOut',
  'animate-openUpRightReturn',
  'animate-perspectiveDown',
  'animate-perspectiveDownReturn',
  'animate-perspectiveLeft',
  'animate-perspectiveLeftReturn',
  'animate-perspectiveRight',
  'animate-perspectiveRightReturn',
  'animate-perspectiveUp',
  'animate-perspectiveUpReturn',
  'animate-ping',
  'animate-puffIn',
  'animate-puffOut',
  'animate-pulse',
  'animate-pulse-alt',
  'animate-roll-in',
  'animate-roll-out',
  'animate-rollIn',
  'animate-rollOut',
  'animate-rotate-in',
  'animate-rotate-in-down-left',
  'animate-rotate-in-down-right',
  'animate-rotate-in-up-left',
  'animate-rotate-in-up-right',
  'animate-rotate-out',
  'animate-rotate-out-down-left',
  'animate-rotate-out-down-right',
  'animate-rotate-out-up-left',
  'animate-rotate-out-up-right',
  'animate-rotateDown',
  'animate-rotateIn',
  'animate-rotateInDownLeft',
  'animate-rotateInDownRight',
  'animate-rotateInUpLeft',
  'animate-rotateInUpRight',
  'animate-rotateLeft',
  'animate-rotateOut',
  'animate-rotateOutDownLeft',
  'animate-rotateOutDownRight',
  'animate-rotateOutUpLeft',
  'animate-rotateOutUpRight',
  'animate-rotateRight',
  'animate-rotateUp',
  'animate-rubber-band',
  'animate-rubberBand',
  'animate-shake',
  'animate-shake-x',
  'animate-shake-y',
  'animate-shakeX',
  'animate-shakeY',
  'animate-slide-in-down',
  'animate-slide-in-left',
  'animate-slide-in-right',
  'animate-slide-in-up',
  'animate-slide-out-down',
  'animate-slide-out-left',
  'animate-slide-out-right',
  'animate-slide-out-up',
  'animate-slideDown',
  'animate-slideDownReturn',
  'animate-slideInDown',
  'animate-slideInLeft',
  'animate-slideInRight',
  'animate-slideInUp',
  'animate-slideLeft',
  'animate-slideLeftReturn',
  'animate-slideOutDown',
  'animate-slideOutLeft',
  'animate-slideOutRight',
  'animate-slideOutUp',
  'animate-slideRight',
  'animate-slideRightReturn',
  'animate-slideUp',
  'animate-slideUpReturn',
  'animate-spaceInDown',
  'animate-spaceInLeft',
  'animate-spaceInRight',
  'animate-spaceInUp',
  'animate-spaceOutDown',
  'animate-spaceOutLeft',
  'animate-spaceOutRight',
  'animate-spaceOutUp',
  'animate-spin',
  'animate-swap',
  'animate-swashIn',
  'animate-swashOut',
  'animate-swing',
  'animate-tada',
  'animate-tinDownIn',
  'animate-tinDownOut',
  'animate-tinLeftIn',
  'animate-tinLeftOut',
  'animate-tinRightIn',
  'animate-tinRightOut',
  'animate-tinUpIn',
  'animate-tinUpOut',
  'animate-twisterInDown',
  'animate-twisterInUp',
  'animate-vanishIn',
  'animate-vanishOut',
  'animate-wobble',
  'animate-zoom-in',
  'animate-zoom-in-down',
  'animate-zoom-in-left',
  'animate-zoom-in-right',
  'animate-zoom-in-up',
  'animate-zoom-out',
  'animate-zoom-out-down',
  'animate-zoom-out-left',
  'animate-zoom-out-right',
  'animate-zoom-out-up',
  'animate-zoomIn',
  'animate-zoomInDown',
  'animate-zoomInLeft',
  'animate-zoomInRight',
  'animate-zoomInUp',
  'animate-zoomOut',
  'animate-zoomOutDown',
  'animate-zoomOutLeft',
  'animate-zoomOutRight',
  'animate-zoomOutUp',
];

const isAnimating = ref(false);
const animationName = ref();

async function setAnimated(value: string = animationName.value) {
  if (isAnimating.value) {
    animationName.value = undefined;
    isAnimating.value = false;

    await nextTick();
  }

  animationName.value = value;
  isAnimating.value = true;
}

function setAnimatedStop() {
  isAnimating.value = false;
}
</script>

<template>
  <div class="relative grid grid-cols-[1fr_250px] h-screen gap-4">
    <button
      class="absolute left-10px top-10px"
      @click="isAnimating ? setAnimatedStop() : setAnimated()"
    >
      play
    </button>

    <div
      class="size-25 flex animate-count-2 items-center self-center justify-center justify-self-center rounded bg-sky text-center font-bold"
      :class="[
        {
          [animationName]: isAnimating,
        },
      ]"
      @animationend="setAnimatedStop"
    >
      Animated
    </div>

    <div class="overflow-auto p-9">
      <div class="flex flex-col gap-2">
        <button
          v-for="item in classLists"
          :key="item"
          @click="setAnimated(item)"
        >
          {{ item.replace('animate-', '') }}
        </button>
      </div>
    </div>
  </div>
</template>

<style lang="postcss">
body {
  margin: 0;
}
</style>
