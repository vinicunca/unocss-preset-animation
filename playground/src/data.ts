// @unocss-include

export const animationNames = [
  'animate-back-in-down',
  'animate-back-in-left',
  'animate-back-in-right',
  'animate-back-in-up',
  'animate-back-out-down',
  'animate-back-out-left',
  'animate-back-out-right',
  'animate-back-out-up',
  'animate-boing-in-up',
  'animate-boing-out-down',
  'animate-bomb-left-out',
  'animate-bomb-right-out',
  'animate-bounce',
  'animate-bounce-alt',
  'animate-bounce-in',
  'animate-bounce-in-down',
  'animate-bounce-in-left',
  'animate-bounce-in-right',
  'animate-bounce-in-up',
  'animate-bounce-out',
  'animate-bounce-out-down',
  'animate-bounce-out-left',
  'animate-bounce-out-right',
  'animate-bounce-out-up',
  'animate-fade-in',
  'animate-fade-in-bottom-left',
  'animate-fade-in-bottom-right',
  'animate-fade-in-down',
  'animate-fade-in-down-big',
  'animate-fade-in-left',
  'animate-fade-in-left-big',
  'animate-fade-in-right',
  'animate-fade-in-right-big',
  'animate-fade-in-top-left',
  'animate-fade-in-top-right',
  'animate-fade-in-up',
  'animate-fade-in-up-big',
  'animate-fade-out',
  'animate-fade-out-bottom-left',
  'animate-fade-out-bottom-right',
  'animate-fade-out-down',
  'animate-fade-out-down-big',
  'animate-fade-out-left',
  'animate-fade-out-left-big',
  'animate-fade-out-right',
  'animate-fade-out-right-big',
  'animate-fade-out-top-left',
  'animate-fade-out-top-right',
  'animate-fade-out-up',
  'animate-fade-out-up-big',
  'animate-flash',
  'animate-flip',
  'animate-flip-in-x',
  'animate-flip-in-y',
  'animate-flip-out-x',
  'animate-flip-out-y',
  'animate-foolish-in',
  'animate-foolish-out',
  'animate-head-shake',
  'animate-heart-beat',
  'animate-hinge',
  'animate-hole-out',
  'animate-jack-in-the-box',
  'animate-jello',
  'animate-light-speed-in-left',
  'animate-light-speed-in-right',
  'animate-light-speed-out-left',
  'animate-light-speed-out-right',
  'animate-magic',
  'animate-open-down-left',
  'animate-open-down-left-out',
  'animate-open-down-left-return',
  'animate-open-down-right',
  'animate-open-down-right-out',
  'animate-open-down-right-return',
  'animate-open-up-left',
  'animate-open-up-left-out',
  'animate-open-up-left-return',
  'animate-open-up-right',
  'animate-open-up-right-out',
  'animate-open-up-right-return',
  'animate-perspective-down',
  'animate-perspective-down-return',
  'animate-perspective-left',
  'animate-perspective-left-return',
  'animate-perspective-right',
  'animate-perspective-right-return',
  'animate-perspective-up',
  'animate-perspective-up-return',
  'animate-ping',
  'animate-puff-in',
  'animate-puff-out',
  'animate-pulse',
  'animate-pulse-alt',
  'animate-roll-in',
  'animate-roll-out',
  'animate-rotate-down',
  'animate-rotate-in',
  'animate-rotate-in-down-left',
  'animate-rotate-in-down-right',
  'animate-rotate-in-up-left',
  'animate-rotate-in-up-right',
  'animate-rotate-left',
  'animate-rotate-out',
  'animate-rotate-out-down-left',
  'animate-rotate-out-down-right',
  'animate-rotate-out-up-left',
  'animate-rotate-out-up-right',
  'animate-rotate-right',
  'animate-rotate-up',
  'animate-rubber-band',
  'animate-shake',
  'animate-shake-x',
  'animate-shake-y',
  'animate-slide-down',
  'animate-slide-down-return',
  'animate-slide-in-down',
  'animate-slide-in-left',
  'animate-slide-in-right',
  'animate-slide-in-up',
  'animate-slide-left',
  'animate-slide-left-return',
  'animate-slide-out-down',
  'animate-slide-out-left',
  'animate-slide-out-right',
  'animate-slide-out-up',
  'animate-slide-right',
  'animate-slide-right-return',
  'animate-slide-up',
  'animate-slide-up-return',
  'animate-space-in-down',
  'animate-space-in-left',
  'animate-space-in-right',
  'animate-space-in-up',
  'animate-space-out-down',
  'animate-space-out-left',
  'animate-space-out-right',
  'animate-space-out-up',
  'animate-spin',
  'animate-swap',
  'animate-swash-in',
  'animate-swash-out',
  'animate-swing',
  'animate-tada',
  'animate-tin-down-in',
  'animate-tin-down-out',
  'animate-tin-left-in',
  'animate-tin-left-out',
  'animate-tin-right-in',
  'animate-tin-right-out',
  'animate-tin-up-in',
  'animate-tin-up-out',
  'animate-twister-in-down',
  'animate-twister-in-up',
  'animate-vanish-in',
  'animate-vanish-out',
  'animate-wobble',
  'animate-zoom-in',
  'animate-zoom-in-down',
  'animate-zoom-in-left',
  'animate-zoom-in-right',
  'animate-zoom-in-up',
  'animate-zoom-out',
  'animate-zoom-out-down',
  'animate-zoom-out-left',
  'animate-zoom-out-right',
  'animate-zoom-out-up',
];
