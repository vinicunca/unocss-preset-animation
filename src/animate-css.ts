export const animateCss = `
@keyframes bounce {

  0%,
  20%,
  53%,
  to {
    animation-timing-function: cubic-bezier(.215, .61, .355, 1);
    transform: translateZ(0)
  }

  40%,
  43% {
    animation-timing-function: cubic-bezier(.755, .05, .855, .06);
    transform: translate3d(0, -30px, 0) scaleY(1.1)
  }

  70% {
    animation-timing-function: cubic-bezier(.755, .05, .855, .06);
    transform: translate3d(0, -15px, 0) scaleY(1.05)
  }

  80% {
    transition-timing-function: cubic-bezier(.215, .61, .355, 1);
    transform: translateZ(0) scaleY(.95)
  }

  90% {
    transform: translate3d(0, -4px, 0) scaleY(1.02)
  }
}

.animate__bounce {
  animation-name: bounce;
  transform-origin: center bottom;
}
`;
