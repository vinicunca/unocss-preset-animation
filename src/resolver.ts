import type { DeepPartial, Preset } from 'unocss';

import * as P from '@vinicunca/perkakas';
import { presetIcons, presetUno, presetWebFonts } from 'unocss';

import type { FluidOptions, ResolvedOptions, VinicuncaOptions } from './types';

import { getAnimationTheme } from './core/animation.theme';

export const defaultFluidOptions: FluidOptions = {
  maxWidth: 1440,
  minWidth: 375,
  remBase: 16,
  useRemByDefault: false,
  extendMaxWidth: null,
  extendMinWidth: null,
  ranges: null,
  commentHelpers: false,
};

export const defaultOptions: VinicuncaOptions = {
  uno: true,
  icons: true,
  webFonts: false,
  enableAnimations: true,
  enableResetStyles: true,
  fluidOptions: defaultFluidOptions,
};

export function resolveOptions(options: DeepPartial<VinicuncaOptions>): ResolvedOptions {
  const optionsWithDefault = P.mergeDeep(defaultOptions, options) as VinicuncaOptions;

  const presets: Array<Preset> = [];
  const presetMap = {
    uno: presetUno,
    icons: presetIcons,
    webFonts: presetWebFonts,
  };

  for (const [key, preset] of Object.entries(presetMap)) {
    const option = optionsWithDefault[key as keyof typeof presetMap];

    if (option) {
      const preset_: any = preset;

      presets.push(
        preset_(
          P.isBoolean(option)
            ? {}
            : option,
        ),
      );
    }
  }

  const animationTheme = optionsWithDefault.enableAnimations
    ? getAnimationTheme()
    : {};

  return {
    ...optionsWithDefault,
    animationTheme,
    presets,
  };
}
