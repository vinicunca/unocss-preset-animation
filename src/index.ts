import type { CSSObject, Preset } from 'unocss';

import { toKebabCase } from '@vinicunca/perkakas';
import postcss from 'postcss';
import { objectify, parse } from 'postcss-js';

import { animateCss } from './animate-css';

type CSSObjectEntries = Array<[string, CSSObject]>;

function toCss(obj: any): string {
  return parse(obj).toString();
}

export function presetAnimation(): Preset {
  const root = postcss.parse(animateCss);
  const obj = objectify(root);

  const properties: CSSObjectEntries = [];

  const keyframes: Array<[string, string]> = [];

  for (const [name, body] of Object.entries(obj)) {
    if (name.startsWith('@keyframes')) {
      keyframes.push([
        name.replace('@keyframes', '').trim(),
        `{${toCss(body)}}`,
      ]);
    } else if (name.startsWith('.animate__')) {
      const attributes: Record<string, string> = {};

      Object.keys(body).forEach((key) => {
        if (key !== 'animationName') {
          attributes[toKebabCase(key)] = body[key];
        }
      });

      properties.push([
        name.replace('.animate__', ''),
        {
          ...attributes,
        },
      ]);
    }
  }

  return {
    name: 'vinicunca-preset-animation',

    layers: {
      animation: -10,
    },

    theme: {
      animation: {
        keyframes: {
          ...Object.fromEntries(keyframes),
        },
        properties: {
          ...Object.fromEntries(properties),
        },
      },
    },
  };
}
