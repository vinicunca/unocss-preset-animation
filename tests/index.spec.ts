import prettier from 'prettier';
import { createGenerator, presetUno } from 'unocss';
import { expect, it } from 'vitest';

// import dedent from 'dedent'
import { presetAnimation } from '../src';

// const prettyCSS = (css: string) => prettier.format(dedent(css), { parser: 'css' })

function prettyCss(css: string) {
  return prettier.format(css, { parser: 'css' });
}

it('preset-animatecss', async () => {
  const uno = createGenerator({
    presets: [
      presetUno({
        preflight: false,
      }),
      presetAnimation(),
    ],
  });

  const root = await uno.generate('animate-bounce');
  const css = await prettyCss(root.css);
  expect(css).toMatchSnapshot();
});
