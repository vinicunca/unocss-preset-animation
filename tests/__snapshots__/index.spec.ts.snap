// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preset-animatecss 1`] = `
"/* layer: default */
@keyframes bounce {
  0%,
  20%,
  53%,
  to {
    animation-timing-function: cubic-bezier(0.215, 0.61, 0.355, 1);
    transform: translateZ(0);
  }
  40%,
  43% {
    animation-timing-function: cubic-bezier(0.755, 0.05, 0.855, 0.06);
    transform: translate3d(0, -30px, 0) scaleY(1.1);
  }
  70% {
    animation-timing-function: cubic-bezier(0.755, 0.05, 0.855, 0.06);
    transform: translate3d(0, -15px, 0) scaleY(1.05);
  }
  80% {
    transition-timing-function: cubic-bezier(0.215, 0.61, 0.355, 1);
    transform: translateZ(0) scaleY(0.95);
  }
  90% {
    transform: translate3d(0, -4px, 0) scaleY(1.02);
  }
}
.animate-bounce {
  animation: bounce 1s linear infinite;
  transform-origin: center bottom;
}
"
`;
